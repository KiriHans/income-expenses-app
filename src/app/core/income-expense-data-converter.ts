import {
  FirestoreDataConverter,
  QueryDocumentSnapshot,
  SnapshotOptions,
} from '@angular/fire/firestore';
import { IncomeExpense } from './models/income-expenses.model';

export const incomeExpenseConverter: FirestoreDataConverter<IncomeExpense> = {
  toFirestore: (incomeExpense: IncomeExpense) => {
    const { description, amount, type, uid } = incomeExpense;
    return { description, amount, type, uid };
  },
  fromFirestore: (
    snapshot: QueryDocumentSnapshot<IncomeExpense>,
    options: SnapshotOptions | undefined
  ) => {
    const data = snapshot.data(options);
    return new IncomeExpense(data.description, data.amount, data.type, data.uid);
  },
};
