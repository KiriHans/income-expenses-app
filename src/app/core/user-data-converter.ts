import {
  FirestoreDataConverter,
  QueryDocumentSnapshot,
  SnapshotOptions,
} from '@angular/fire/firestore';
import { UserApp } from './models/users.model';

export const userConverter: FirestoreDataConverter<UserApp> = {
  toFirestore: (user: UserApp) => {
    const { displayName, email } = user;
    return { displayName, email };
  },
  fromFirestore: (
    snapshot: QueryDocumentSnapshot<UserApp>,
    options: SnapshotOptions | undefined
  ) => {
    const data = snapshot.data(options);
    return new UserApp(data.id, data.displayName, data.email);
  },
};
