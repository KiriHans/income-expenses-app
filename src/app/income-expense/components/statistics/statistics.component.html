<div class="row row-cols-1">
  <div class="col col-md-6 pl-0 pl-sm-3 pb-5 pb-md-0">
    <div class="card card-statistics">
      <div class="card-body">
        <div class="clearfix">
          <div class="float-left">
            <i class="fa fa-money-bill-alt text-success icon-lg"></i>
          </div>
          <div class="float-right">
            <p class="mb-0 text-right">Income</p>
            <div class="fluid-container">
              <h3 class="font-weight-medium text-right mb-0">
                {{ this.totalIncome() | currency }}
              </h3>
            </div>
          </div>
        </div>
        <p class="text-muted mt-3 mb-0">
          <i class="mdi mdi-alert-octagon mr-1" aria-hidden="true"></i>
          {{ this.numberIncomeItems() }} items
        </p>
      </div>
    </div>
  </div>
  <div class="w-100 d-none d-md-none"></div>
  <div class="col col-md-6 pl-0 pl-sm-3 pb-5 pb-md-0">
    <div class="card card-statistics">
      <div class="card-body">
        <div class="clearfix">
          <div class="float-left">
            <i class="fa fa-shipping-fast text-danger icon-lg"></i>
          </div>
          <div class="float-right">
            <p class="mb-0 text-right">Expense</p>
            <div class="fluid-container">
              <h3 class="font-weight-medium text-right mb-0">
                {{ this.totalExpense() | currency }}
              </h3>
            </div>
          </div>
        </div>
        <p class="text-muted mt-3 mb-0">
          <i class="mdi mdi-alert-octagon mr-1" aria-hidden="true"></i>
          {{ this.numberExpenseItems() }} items
        </p>
      </div>
    </div>
  </div>
</div>

<div class="row mt-5">
  <div class="col-12 pl-0 pl-sm-3 pb-5 pb-md-0">
    <div class="card card-statistics">
      <div class="card-body">
        <div class="clearfix">
          <div class="float-left">
            <i class="fa fa-building text-info icon-lg"></i>
          </div>
          <div class="float-right">
            <p class="mb-0 text-right">Difference</p>
            <div class="fluid-container">
              <h3
                [ngClass]="{
                  'text-success': differenceIncome() > 0,
                  'text-danger': differenceIncome() < 0
                }"
                class="font-weight-medium text-right mb-0"
              >
                {{ differenceIncome() | currency }}
              </h3>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>

  <div class="col pl-1 pt-5 p-sm-5" *ngIf="totalIncome() > 0 || totalExpense() > 0">
    <div class="card card-statistics">
      <div class="card-body">
        <h1 class="card-title">Graphic</h1>
        <div style="min-width: 200px" class="w-50 mx-auto">
          <canvas
            baseChart
            class="chart"
            [data]="this.doughnutChartData()"
            [type]="this.doughnutChartType"
          >
          </canvas>
        </div>
      </div>
    </div>
  </div>
</div>
