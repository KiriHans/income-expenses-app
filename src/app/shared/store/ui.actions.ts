import { createActionGroup, emptyProps } from '@ngrx/store';

export const UiActions = createActionGroup({
  source: 'UI',
  events: {
    'Load UI': emptyProps(),
    'Stop Load UI': emptyProps(),
  },
});
