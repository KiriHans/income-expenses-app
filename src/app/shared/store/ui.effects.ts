import { Injectable } from '@angular/core';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { map, concatMap } from 'rxjs/operators';
import { EMPTY } from 'rxjs';
import { UiActions } from './ui.actions';

@Injectable()
export class UiEffects {
  loadUis$ = createEffect(() => {
    return this.actions$.pipe(
      ofType(UiActions.loadUI),
      concatMap(() =>
        /** An EMPTY observable only emits completion. Replace with your own observable API request */
        EMPTY.pipe(map((data) => UiActions.loadUI()))
      )
    );
  });

  constructor(private actions$: Actions) {}
}
