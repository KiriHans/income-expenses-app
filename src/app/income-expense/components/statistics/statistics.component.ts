import { Component, Signal, computed } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Store } from '@ngrx/store';
import {
  selectNumberExpenseItems,
  selectNumberIncomeItems,
  selectTotalExpense,
  selectTotalIncome,
} from 'src/app/store/selectors/income-expense.selectors';
import { ChartData, ChartType } from 'chart.js';
import { NgChartsModule } from 'ng2-charts';

@Component({
  selector: 'app-statistics',
  standalone: true,
  imports: [CommonModule, NgChartsModule],
  templateUrl: './statistics.component.html',
  styles: [],
})
export class StatisticsComponent {
  numberIncomeItems!: Signal<number>;
  numberExpenseItems!: Signal<number>;
  totalIncome!: Signal<number>;
  totalExpense!: Signal<number>;

  differenceIncome!: Signal<number>;

  doughnutChartLabels: string[] = ['Income', 'Expense'];
  doughnutChartData!: Signal<ChartData<'doughnut'>>;
  doughnutChartType: ChartType = 'doughnut';

  constructor(private store: Store) {
    this.generateStatistics();
  }

  generateStatistics(): void {
    this.numberIncomeItems = this.store.selectSignal(selectNumberExpenseItems);
    this.numberExpenseItems = this.store.selectSignal(selectNumberIncomeItems);
    this.totalIncome = this.store.selectSignal(selectTotalIncome);
    this.totalExpense = this.store.selectSignal(selectTotalExpense);

    this.differenceIncome = computed(() => this.totalIncome() - this.totalExpense());

    this.doughnutChartData = computed(() => ({
      labels: this.doughnutChartLabels,
      datasets: [
        {
          data: [this.totalIncome(), this.totalExpense()],
          hoverOffset: 4,
          backgroundColor: ['rgb(75, 192, 192)', 'rgb(255, 99, 132)'],
        },
      ],
    }));
  }
}
