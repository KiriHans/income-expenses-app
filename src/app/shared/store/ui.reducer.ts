import { createFeature, createReducer, on } from '@ngrx/store';
import { UiActions } from './ui.actions';

export const uiFeatureKey = 'ui';

export interface UIState {
  isLoading: boolean;
}

export const initialState: UIState = {
  isLoading: false,
};

export const uiReducer = createReducer(
  initialState,
  on(UiActions.loadUI, (state): UIState => ({ ...state, isLoading: true })),
  on(UiActions.stopLoadUI, (state): UIState => ({ ...state, isLoading: false }))
);

export const uiFeature = createFeature({
  name: uiFeatureKey,
  reducer: uiReducer,
});
