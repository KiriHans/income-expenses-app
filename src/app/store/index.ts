import { isDevMode } from '@angular/core';
import { ActionReducerMap, MetaReducer } from '@ngrx/store';
import * as ui from '../shared/store/ui.reducer';
import * as auth from './reducers/auth.reducer';
import * as incomeExpense from './reducers/income-expense.reducer';

export const stateFeatureKey = 'state';

export interface AppState {
  [ui.uiFeatureKey]: ui.UIState;
  [auth.authFeatureKey]: auth.State;
  [incomeExpense.incomeExpenseFeatureKey]: incomeExpense.State;
}

export const reducers: ActionReducerMap<AppState> = {
  [ui.uiFeatureKey]: ui.uiFeature.reducer,
  [auth.authFeatureKey]: auth.userReducer,
  [incomeExpense.incomeExpenseFeatureKey]: incomeExpense.reducer,
};

export const metaReducers: MetaReducer<AppState>[] = isDevMode() ? [] : [];
