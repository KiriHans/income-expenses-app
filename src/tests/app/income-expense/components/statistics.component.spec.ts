import { CurrencyPipe } from '@angular/common';
import { DebugElement } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';

import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { NgChartsModule } from 'ng2-charts';
import { IncomeExpense } from 'src/app/core/models/income-expenses.model';
import { StatisticsComponent } from 'src/app/income-expense/components/statistics/statistics.component';

describe('StatisticsComponent', () => {
  let component: StatisticsComponent;
  let fixture: ComponentFixture<StatisticsComponent>;
  let el: DebugElement;
  let compiled: HTMLElement;

  let currencyPipe: CurrencyPipe;

  let store: MockStore;

  const initialState = {
    incomeExpense: {
      items: [
        new IncomeExpense('Test 1', 100, 'income'),
        new IncomeExpense('Test 2', 200, 'income'),
        new IncomeExpense('Test 3', 400, 'expense'),
      ],
      loading: false,
    },
  } as const;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [StatisticsComponent, NgChartsModule],
      providers: [provideMockStore({ initialState }), CurrencyPipe],
    });
    fixture = TestBed.createComponent(StatisticsComponent);
    component = fixture.componentInstance;
    el = fixture.debugElement;
    compiled = fixture.nativeElement;

    store = TestBed.inject(MockStore);
    currencyPipe = TestBed.inject(CurrencyPipe);

    window.ResizeObserver = jest.fn(() => ({
      observe: jest.fn(),
      unobserve: jest.fn(),
      disconnect: jest.fn(),
    }));

    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  it('should contain an income card that shows the total income of the user', () => {
    const totalIncome = component.totalIncome();
    const numberItems = component.numberIncomeItems();
    const incomeCard = el.query(By.css('.income-card'));

    expect(incomeCard).toBeTruthy();

    const title = incomeCard.query(By.css('div.float-right p'));
    const totalIncomeCard = incomeCard.query(By.css('h3'));
    const numberItemsCard = incomeCard.query(By.css('p.text-muted'));

    expect(title.nativeElement.textContent).toBe('Income');
    expect(totalIncomeCard.nativeElement.textContent.trim()).toBe(
      `${currencyPipe.transform(totalIncome)}`
    );
    expect(numberItemsCard.nativeElement.textContent.trim()).toBe(`${numberItems} items`);
  });

  it('should contain an expense card that shows the total expense of the user', () => {
    const totalExpense = component.totalExpense();
    const numberItems = component.numberExpenseItems();
    const expenseCard = el.query(By.css('.expense-card'));

    expect(expenseCard).toBeTruthy();

    const title = expenseCard.query(By.css('div.float-right p'));
    const totalExpenseCard = expenseCard.query(By.css('h3'));
    const numberItemsCard = expenseCard.query(By.css('p.text-muted'));

    expect(title.nativeElement.textContent).toBe('Expense');
    expect(totalExpenseCard.nativeElement.textContent.trim()).toBe(
      `${currencyPipe.transform(totalExpense)}`
    );
    expect(numberItemsCard.nativeElement.textContent.trim()).toBe(`${numberItems} items`);
  });

  it('should contain an difference card that shows the total difference of the user', () => {
    const differenceIncome = component.differenceIncome();
    const differenceCard = el.query(By.css('.difference-card'));

    expect(differenceCard).toBeTruthy();

    const title = differenceCard.query(By.css('div.float-right p'));
    const totalDifferenceCard = differenceCard.query(By.css('h3'));

    expect(title.nativeElement.textContent).toBe('Difference');
    expect(totalDifferenceCard.nativeElement.textContent.trim()).toBe(
      `${currencyPipe.transform(differenceIncome)}`
    );
  });

  it('Total difference should have its respective class depending if the difference is positive or not', () => {
    let totalDifferenceCard = el.query(By.css('.difference-card h3'));

    expect(totalDifferenceCard.classes['text-danger']).toBeTruthy();

    store.setState({
      incomeExpense: {
        items: [
          new IncomeExpense('Test 1', 100, 'income'),
          new IncomeExpense('Test 2', 200, 'income'),
          new IncomeExpense('Test 3', 150, 'expense'),
        ],
        loading: false,
      },
    });

    store.refreshState();

    fixture.detectChanges();

    totalDifferenceCard = el.query(By.css('.difference-card h3'));

    expect(totalDifferenceCard.classes['text-success']).toBeTruthy();
  });

  it('should contain a graphic that shows the data', () => {
    const chartCard = el.query(By.css('.graphic-card'));

    expect(chartCard).toBeTruthy();

    const title = chartCard.query(By.css('h1'));
    const chart = chartCard.query(By.css(`canvas`));

    expect(title.nativeElement.textContent).toBe('Graphic');
    expect(chart.classes['chart']).toBeTruthy();
  });

  it('should generate a correct dataset depending of the user income-expense', () => {
    const totalIncome = component.totalIncome();
    const totalExpense = component.totalExpense();
    const charData = component.doughnutChartData();
    const dataset = charData.datasets[0];

    expect(dataset.data).toEqual([totalIncome, totalExpense]);
  });
});
