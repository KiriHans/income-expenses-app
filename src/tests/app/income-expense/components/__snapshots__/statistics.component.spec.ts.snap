// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticsComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="row row-cols-1"
  >
    <div
      class="income-card col col-md-6 pl-0 pl-sm-3 pb-5 pb-md-0"
    >
      <div
        class="card card-statistics"
      >
        <div
          class="card-body"
        >
          <div
            class="clearfix"
          >
            <div
              class="float-left"
            >
              <i
                class="fa fa-money-bill-alt text-success icon-lg"
              />
            </div>
            <div
              class="float-right"
            >
              <p
                class="mb-0 text-right"
              >
                Income
              </p>
              <div
                class="fluid-container"
              >
                <h3
                  class="font-weight-medium text-right mb-0"
                >
                   $300.00 
                </h3>
              </div>
            </div>
          </div>
          <p
            class="text-muted mt-3 mb-0"
          >
            <i
              aria-hidden="true"
              class="mdi mdi-alert-octagon mr-1"
            />
             1 items 
          </p>
        </div>
      </div>
    </div>
    <div
      class="w-100 d-none d-md-none"
    />
    <div
      class="expense-card col col-md-6 pl-0 pl-sm-3 pb-5 pb-md-0"
    >
      <div
        class="card card-statistics"
      >
        <div
          class="card-body"
        >
          <div
            class="clearfix"
          >
            <div
              class="float-left"
            >
              <i
                class="fa fa-shipping-fast text-danger icon-lg"
              />
            </div>
            <div
              class="float-right"
            >
              <p
                class="mb-0 text-right"
              >
                Expense
              </p>
              <div
                class="fluid-container"
              >
                <h3
                  class="font-weight-medium text-right mb-0"
                >
                   $400.00 
                </h3>
              </div>
            </div>
          </div>
          <p
            class="text-muted mt-3 mb-0"
          >
            <i
              aria-hidden="true"
              class="mdi mdi-alert-octagon mr-1"
            />
             2 items 
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="row mt-5"
  >
    <div
      class="difference-card col-12 pl-0 pl-sm-3 pb-5 pb-md-0"
    >
      <div
        class="card card-statistics"
      >
        <div
          class="card-body"
        >
          <div
            class="clearfix"
          >
            <div
              class="float-left"
            >
              <i
                class="fa fa-building text-info icon-lg"
              />
            </div>
            <div
              class="float-right"
            >
              <p
                class="mb-0 text-right"
              >
                Difference
              </p>
              <div
                class="fluid-container"
              >
                <h3
                  class="font-weight-medium text-right mb-0 text-danger"
                >
                   -$100.00 
                </h3>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="graphic-card col pl-1 pt-5 p-sm-5"
    >
      <div
        class="card card-statistics"
      >
        <div
          class="card-body"
        >
          <h1
            class="card-title"
          >
            Graphic
          </h1>
          <div
            class="w-50 mx-auto"
            style="min-width: 200px;"
          >
            <canvas
              basechart=""
              class="chart"
              height="0"
              style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
              width="0"
            />
          </div>
        </div>
      </div>
    </div>
    
  </div>
</div>
`;
